File: app/Quotes201812/c/src/servers.c
// Copyright 11-Sept-2018 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "servers.h"
#include "data/servers_db.h"
#include "data/nicks_db.h"
#include "conf.h"

CgiRp *servers_process(Mjson *rqm) {
  char *rq = jmap_gstring(rqm, "rq");

  // -------------------------------------------------------------- idata
  if (str_eq(rq, "idata")) {
    Achar *servers = achar_new();
    EACH(servers_db_list(), Server, s) {
      achar_add(servers, server_name(s));
    }_EACH

    Mjson *conf = json_robject(conf_get());
    char *selServer = jmap_gstring(conf, "server_id");
    if (!*selServer) {
      selServer = achar_get(servers, 0);
    }

    Anick *nicks = nicks_db_list();

    Mjson *m = mjson_new();
    jmap_pstring(m, "selServer", selServer);
    mjson_put(m, "servers", achar_to_json(servers));
    mjson_put(m, "nicks", anick_to_json(nicks));
    return cgi_ok(m);
  }

  // ------------------------------------------------------- nickCode
  if (str_eq(rq, "nickCode")) {
    char *serverId = jmap_gstring(rqm, "serverId");
    char *nickId = jmap_gstring(rqm, "nickId");
    Server *sv = servers_db_get(serverId);
    Mchar *nicks = server_nicks(sv);
    char *code = mchar_oget(nicks, nickId, "");
    Mjson *m = mjson_new();
    jmap_pstring(m, "code", code);
    return cgi_ok(m);
  }

  // ------------------------------------------------------- testCode
  if (str_eq(rq, "testCode")) {
    char *serverId = jmap_gstring(rqm, "serverId");
    char *code = jmap_gstring(rqm, "code");
    Server *sv = servers_db_get(serverId);
    bool ok = true;
    if (oaquote_is_null(server_read(sv)(code))) {
      ok = false;
    }
    Mjson *m = mjson_new();
    jmap_pbool(m, "ok", ok);
    return cgi_ok(m);
  }

  // ------------------------------------------------------- testCode
  if (str_eq(rq, "setCode")) {
    char *serverId = jmap_gstring(rqm, "serverId");
    char *nickId = jmap_gstring(rqm, "nickId");
    char *code = jmap_gstring(rqm, "code");
    Server *sv = servers_db_get(serverId);
    Mchar *codes = server_nicks(sv);
    mchar_put(codes, nickId, code);
    server_set_nicks(sv, codes);
    return cgi_ok(mjson_new());
  }

  // ------------------------------------------------------- setSelServer
  if (str_eq(rq, "setSelServer")) {
    char *serverId = jmap_gstring(rqm, "name");
    conf_set_server_id(serverId);
    return cgi_ok(mjson_new());
  }

  THROW("") "'%s': Unknown request in servers", rq _THROW
  // Unreacheable
  return NULL;

}

File: app/Quotes/h/src/Servers.hs
-- Copyright 03-Dic-2018 ºDeme
-- GNU General Public License - V3 <http://www.gnu.org/licenses/>

-- | Servers page

module Servers (process) where

import Data.List
import Text.Printf
import qualified Dm.Js as Js
import Dm.Js (JSValue)
import qualified Dm.Cgi as Cgi
import Dm.Cgi (Cgi)
import qualified Data.NicksDb as NicksDb
import qualified Data.ServersDb as ServersDb
import qualified Data.Server as Server
import qualified Conf as Conf

process :: Cgi -> [(String, JSValue)] -> IO ()
process cgi rq =
  case Cgi.get rq Js.rString "rq" of
    "idata" -> do ------------------------------------------------------- idata
      nicks <- NicksDb.readJs
      let (s1, s2, s3) = ServersDb.list
      let servers = map (Js.wString) $ ServersDb.listNames
      selServer' <- Conf.getValue "serverId"
      let selServer = if Js.rString selServer' == "" then head servers
                                                     else selServer'
      Cgi.ok cgi [("selServer", selServer),
                  ("servers", Js.wList servers),
                  ("nicks", nicks !! 2)]
    "nickCode" ->  do ------------------------------------------------ nickCode
      let serverId = Cgi.get rq Js.rString "serverId"
      let nickId = Cgi.get rq Js.rString "nickId"
      codes <- ServersDb.nicks serverId
      case lookup nickId codes of
        Nothing -> Cgi.ok cgi [("code", Js.wString "")]
        Just v -> Cgi.ok cgi [("code", Js.wString v)]
    "setSelServer" -> do  ---------------------------------------- setSelServer
      let serverId = Cgi.get rq Js.rString "name"
      Conf.set "serverId" $ Js.wString serverId
      Cgi.empty cgi
    s -> error $ printf "Unknown rq '%s'" s ----------------------------- Error
